examples/advanced/main.go

```diff
@@ -36,19 +38,44 @@ import (
 // main is compiled to an exported Wasm function named "_start", called by the
 // Wasm scheduler plugin during initialization.
 func main() {
+	// Instead of using `plugin.Set`, this configures only the interfaces
+	// implemented by the plugin. The Wasm host only calls functions imported,
+	// so this prevents additional overhead.
+	enqueue.SetPlugin(func(h handleapi.Handle) api.EnqueueExtensions {
+		p := pluginInitializer(h)
+		plugin, ok := p.(api.EnqueueExtensions)
+		if ok {
+			return plugin
+		}
+		panic("pluginInitializer did not return an api.EnqueueExtensions")
+	})
+	prescore.SetPlugin(func(h handleapi.Handle) api.PreScorePlugin {
+		p := pluginInitializer(h)
+		plugin, ok := p.(api.PreScorePlugin)
+		if ok {
+			return plugin
+		}
+		panic("pluginInitializer did not return an api.PreScorePlugin")
+	})
+	score.SetPlugin(func(h handleapi.Handle) api.ScorePlugin {
+		p := pluginInitializer(h)
+		plugin, ok := p.(api.ScorePlugin)
+		if ok {
+			return plugin
+		}
+		panic("pluginInitializer did not return an api.ScorePlugin")
+	})
+}
+
+func pluginInitializer(h handleapi.Handle) api.Plugin {
 	// The plugin package uses only normal Go code, which allows it to be
 	// unit testable via `tinygo test -target=wasi` as well normal `go test`.
 	//
 	// The real implementations, such as `config.Get()` use Wasm host functions
 	// (go:wasmimport), which cannot be tested with `tinygo test -target=wasi`.
-	plugin, err := plugin.New(klog.Get(), config.Get())
+	plugin, err := plugin.New(klog.Get(), config.Get(), h)
 	if err != nil {
 		panic(err)
 	}
-	// Instead of using `plugin.Set`, this configures only the interfaces
-	// implemented by the plugin. The Wasm host only calls functions imported,
-	// so this prevents additional overhead.
-	enqueue.SetPlugin(plugin)
-	prescore.SetPlugin(plugin)
-	score.SetPlugin(plugin)
+	return plugin
 }
```

Comment:
It doesn't look great. 

We have two alternative choices:
- change `Set` functions to accept `func (klog klog.Klog, jsonConfig []byte, h handleapi.Handle)`
- change `New` function of this plugin to only have `h handleapi.Handle` parameter, and change it to do klog.Get() and config.Get() in it.

I prefer the first option. Actually, the scheduling framework's initializer type contains configuration in the second parameters, so it's not weird to add it to our plugin initializer type as well. 
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/registry.go#L31
Also, for logger, we may want to add it to handleapi.Handle.

Reply:
Sure, therefore I will change pluginInitializer to  be `func (klog klog.Klog, jsonConfig []byte, h handleapi.Handle)`.

> Also, for logger, we may want to add it to handleapi.Handle.

Then in the near future, we might remove `klog` and integrate it into `handle`.